### ISSUE TITLE
RavenDB strategy

### ISSUE THREAD

**Issue by DavidRieman** [WheelMud/WheelMUD#43]
To be replaced?  Which strategy do we want to start with instead?  (Forking discussion from #33)

**Comment by DavidRieman** [comment]
> why the choice to go with RavenDB was made?

There's a (poorly-formatted) backup copy of one of those conversations at http://muds.gamedev.com/mudforum/index.php?t=msg&goto=257&rid=0

Also other conversations can be found, searching RavenDB
http://muds.gamedev.com/mudforum/index.php?t=msg&goto=420

But I'm pretty sure I'm just not finding the conversation where we talked about not having to add DB columns for every time someone adds a new Behavior or adds a property to one, etc. In short, we want new game-making hobbyists to feel empowered to make cool components, without having to do tons of ceremony around threading data through the DAL; without having to write complex DB-upgrade scripts and the like, and having to take the game offline while they run said upgrade scripts against a database, etc.

For example, if I decide to add an int "hardness" to a weapon behavior, I should be able to give it a reasonable default value (say, 10), and if it passes testing (including automatic persistence tests), drop the new lib into our deployed app for our no-reboot-update process to pick up and upgrade existing weapons behaviors (which get hardness 10 since they didn't have that property before). Now I can immediately start using the online world editor to assign custom hardness values to different weapon drops. Or put in code that starts using the hardness.

**Comment by DavidRieman** [comment]
While one could use a key-to-generic-data-blob solution to stuff arbitrarily serialized data into the generic data column, a document DB theoretically has a lot of benefits over this, like being able to actually peruse the data in a sane manner, or write queries against arbitrary properties in there.  Again I haven't done all this stuff before, but these are the promises of document DB I've heard.

**Comment by pha3z** [comment]
Are you familiar with CouchDB?  My partner and I were looking at it for a project we were entertaining.  The project hasn't gone through, yet, so I haven't looked into it much.  But it was highly recommended to us by someone who works with cloud-hosted solutions and changing data problems like you described.

I am actually presently designing my own MUD backend.  I am working strictly on data structures and world persistence.  The way I see it, getting data figured out should be done before ever writing a single line of code. I think if more developers would learn how to be DBAs, the software we have would be dramatically improved.

This looks very much like the first project I've done that will truly _necessitate_ a nosql document-style database. 

The ORM problem is obviously a hard problem, and is magnified tenfold when you introduce changing data during development.  I haven't tried Entity framework.  I have actually implemented my own style of abstraction in practice, that I personally feel works better than using someone else's ORM.  The interesting thing about the way I ended up doing things is that it seemed to work very well as a middle ground for dealing with the ceremony you talk about.  But, I digress, it would still be too much for changes that are happening as frequently as they would be in a developing MUD.

I saw on the link you sent that one of the original missions of the MUD was to also be a learning platform for real-world coding.  Since the vast majority of business code is against sql relational databases, that was a guiding decision.  To me that seems like trying to stick a round peg in a square hole.  MUD data is extremely dynamic and very far removed from what you will encounter in a business environment.  As such, a document DB makes sense.

I'm holding my breathe until I get time to research CouchDB and specifically research the potential pitfalls of document DB.  But that's the direction that seems most logical to my knowledge.

**Comment by DavidRieman** [comment]
Hmm, a search for CouchDB in the backup forum only turned up one hit.  It's also interesting reading on this subject and probably does a better job explaining why we chose RavenDB than the other links actually: http://muds.gamedev.com/mudforum/index.php?t=msg&goto=382&rid=0

I don't recall for sure, but I feel like my attempted spike into persistence with CouchDB didn't get far.  Like, if we had to serialize into JSON from .NET to send to CouchDB, that could have been my hangup on that one.  Maybe we would have chosen CouchDB if we were a JavaScript stack.

> designing my own MUD backend. I am working strictly on data structures and world persistence

Cool!  If you are interested in helping shape and revitalize our world persistence strategy, perhaps this could be a great fit?  Our big gap in that department has certainly been one of the primary factors in being unmotivated and not getting much done here these last few years.  No pressure though.  Even just these conversations have been helpful.

> I'm holding my breathe

I'm not sure what you mean with this line - do you mean "I'm not 100% which tech is the right one" yet?

**Comment by pha3z** [comment]
On holding my breathe:
Correct. I'm not 100% certain yet which is the right tech.

The conversations we are having sparked my interest to look into databases some while I'm working on data structures and persistence.

I looked up CouchDb briefly and it was mentioned that it's "eventually consistent".  They're really pushing it for extremely wide horizontal scaling.  It doesn't appear to support immediate transactions.

MongoDB, on the other hand, supports immediate multi-document transactions.  That would be the type of thing that is an absolute must for a MUD.  In a MUD world, having everything always up-to-date so you don't get weird bugs (like the item duplication you described) is probably far more important than worrying about massive wide scaling... unless a person has got it in their mind that they're building a mud that's going to get 10,000+ concurrent users. :P

I'm going to make the need for immediate transactions my primary research target, as that will filter out a lot of Document databases.  Not all document databases are designed for extreme wide scale.  However, it's a fact that the primary driver that caused document databases to jump into wide use in the last decade is due to them lending themselves to wide scaling, where relational databases started failing miserably at keeping up with the demands of today's gigantuous user services.

I did a study on ACID in depth many months ago, because I'd been entertaining document databases at that time...

For your own learning, in case you don't know:
One of the big things about document databases is that they often choose to forgo at least one of the ACID principles in order to achieve the mass parallelism that they get.  "Eventual consistency" is an example of partial violation the "Consistency" principle but it makes building wide distributions easy.

In the case of a MUD, I really think you'd want to guarantee all four components of ACID pretty much immediately, which helps limit the scope of no document database options.

As far as helping with the MUD, it's interesting to know there's a need there.  I might help contribute some design advice at the least.  But what I can do is limited for a couple reasons:
for one: I'm actually designing something a bit more than a MUD. I'm straddling the fence with an engine that can double as an interactive novel system.  This adds some specific nuances you might not find in a generic MUD engine.
for two: I might actually write it in Rust (which is one of the most amazing languages ever made, IHMO).  I love C# and would absolutely prefer to stay in C# for the sheer ease and simplicity of it, but one of my design decisions is probably going to mandate 100% cross-platform support and the ease of jumping between systems. As such, I'll probably write the core in Rust as libraries and possibly leverage it from C# frontend.

But yeah, I'm big on data structures and thinking about persistence and in-memory representation and all those problems, so I'll try to stay in tune with what's happening here and at least offer thoughts when I can.

**Comment by pha3z** [comment]
This should probably be put in another issue titled "Data Persistence Model" or something to that effect, but I wasn't sure exactly how you're gaming the issues board on this project.  Please more this if it doesn't belong here.

I know flat files have been brought up, and I addressed the problems flat files bring with dynamic data.  But has anyone thought about whether static data might better suited for flat files than an actual database?  The ridiculously ancient but popular CircleMUD 3 engine used flat files for Zones definitions.  The files were split into rooms, objects, and mobiles.  The format worked well, and there have been graphical world-builders created to read and write the format so builders can work in a graphical editor.

What are the downsides to flat files specifically for static data?

**Comment by DavidRieman** [comment]
I don’t have long ATM but I will be back. 

I think it might be acceptable for a MUD to have less than immediately guaranteed persistence, as long as the right transactions are in place to prevent duplication issues, and the right tech in place to detect when unique instances double up anyways. 

We have a MonoDevelop solution as we once had a dev working in that, but I nearly deleted it. I would be down with continuing maintenance if we had anyone using it. I think our current techs, sqlite and ravendb, are cross-platform for all three. 

We are trying to build in layers so even for example if someone only wanted the telnet protocol handling, we would still be saving them time.

**Comment by DavidRieman** [comment]
> wasn't sure exactly how you're gaming the issues board

I am still working it out. :)

> whether static data might better suited for flat files than an actual database? 

I don’t want to assume any particular behavior to be static. I think this part of the last forum backup I linked demonstrates well enough:

> I don't think it would make sense to store different thing types apart. Things get really hard to reason about when you have crossovers / compicated Things. Imagine for instance a MUD about space combat, where there are no 'player' avatars but instead you are basically a vehicle; some players act as Rooms, having large hanger bays which can house other players, etc... Basically a typical Thing that a player controls has PlayerBehavior, UserControlledBehavior, MovableBehavior, RoomBehavior, etc... Does that entity get stored as a room, as a player, or as both?

Maybe a player as capital ship might even be essentially a zone including multiple hangers etc. (In such a game, players probably typically navigate with coordinates and such over time rather than compass directions through room exits)

Other examples may be procedurally generated world the users can change, player housing zones, a maze that changes all exits with a maze making algorithm whenever a team has held the objective room for long enough, OLC edits of a zone being built, etc.  So varying levels of static to dynamic.  But solving dynamic solves static too.

**Comment by pha3z** [comment]
Very insightful.  I agree with your thoughts.  Thanks for sharing!

**Comment by DavidRieman** [comment]
Moved RavenDB behind repository pattern with #66 so one could build a separate backing document repository if they wanted now and switch to using it with an app.config change.  Saving+loading the player is working better than it has in years and I'm feeling good about it  :)

**Comment by DavidRieman** [comment]
Done.  PS.  Also on "how you're gaming the issues board" - got a better idea on this now.  I've moved to maintaining the base repo, where I have the Issues tab to track all issues, Discussions area for non-issue discussions, and Projects tab to organize the issues into distinct groups we can be a bit more agile with.  (Sprints don't really work for our slow pace, but I hope the new organization will help none-the-less, once built up a bit more.)
